### PR TITLE
Iimproved startup

### PR DESCRIPTION
**Pull request by ejfitzgerald**

### CODE REVIEW

File: libs/chain/src/constants.cpp [hunk 1]
```diff
@@ -17,24 +17,101 @@
...
+
+}  // namespace
+
+Digest GetGenesisDigest()
+{
+  Digest digest = genesis_state.Apply([](GenesisState const &state) { return state.digest; });
+
+  if (digest.empty())
+  {
+    throw std::logic_error("Genesis has not been initialised");
```
Comment (n-hutton) [inline libs/chain/src/constants.cpp:47]: 👍

File: libs/constellation/src/constellation.cpp [hunk 2]
```diff
@@ -91,6 +90,35 @@ constexpr char const *LOGGING_NAME = "constellation";
 const std::size_t HTTP_THREADS{4};
 char const *      GENESIS_FILENAME = "genesis_file.json";
 
+class Defer
+{
+public:
+  using Callback = void (Constellation::*)();
+
+  explicit Defer(Constellation *instance, Callback callback)
```
Comment (n-hutton) [inline libs/constellation/src/constellation.cpp:98]: no need for explicit here

File: libs/constellation/src/constellation.cpp [hunk 3]
```diff
@@ -338,69 +360,225 @@ Constellation::Constellation(CertificatePtr const &certificate, Config config)
...
+  // - recover from previous genesis init
+  {
+    std::string const genesis_file_path =
+        cfg_.genesis_file_location.empty() ? GENESIS_FILENAME : cfg_.genesis_file_location;
+
+    GenesisFileCreator creator(*storage_, external_identity_, cfg_.db_prefix);
+
+    genesis_status = creator.LoadFile(genesis_file_path, cfg_.proof_of_stake, params);
+
+    if (genesis_status == GenesisFileCreator::Result::FAILURE)
```
Comment (n-hutton) [inline libs/constellation/src/constellation.cpp:478]: 👍

File: libs/constellation/src/constellation.cpp [hunk 4]
```diff
@@ -338,69 +360,225 @@ Constellation::Constellation(CertificatePtr const &certificate, Config config)
...
+    {
+      FETCH_LOG_ERROR(LOGGING_NAME, "Failed to restore or generate Genesis configuration");
+      return false;
+    }
+  }
+
+  // create the DAG
```
Comment (n-hutton) [inline libs/constellation/src/constellation.cpp:485]: Probably need to clear up the settings.cpp in another PR

File: libs/constellation/src/constellation.cpp [hunk 5]
```diff
@@ -338,69 +360,225 @@ Constellation::Constellation(CertificatePtr const &certificate, Config config)
...
+                               external_identity_->identity());
+
+  if (cfg_.proof_of_stake)
+  {
+    consensus_->SetDefaultStartTime(params.start_time);
+    consensus_->SetMaxCabinetSize(params.cabinet_size);
+
+    if (params.snapshot)
+    {
+      consensus_->Reset(*params.snapshot);
```
Comment (n-hutton) [inline libs/constellation/src/constellation.cpp:536]: I will probably have to change this in my wealth PR - the flow should be basically the same for POS

File: libs/constellation/src/constellation.cpp [hunk 6]
```diff
@@ -691,50 +739,225 @@ bool Constellation::Run(UriSet const &initial_peers, core::WeakRunnable bootstra
...
+  lane_services_.StopExternal();
+
+  ResetItem(synergetic_miner_);
+  ResetItem(dag_service_);
+
+  http_modules_.clear();
+  ResetItem(health_check_module_);
+  ResetItem(http_open_api_module_);
+
+  ResetItem(messenger_api_);
```
Comment (n-hutton) [inline libs/constellation/src/constellation.cpp:801]: Won't this happen anyway? Is this so you can do it in a particular order, or for easier debugging?

File: libs/ledger/include/ledger/storage_unit/fake_storage_unit.hpp [hunk 7]
```diff
@@ -92,8 +92,8 @@ class FakeStorageUnit final : public StorageUnitInterface
   TransactionStore             transaction_store_{};
   StatePtr                     state_{std::make_shared<State>()};
   StateHistory                 state_history_{};
-  StateHashStack               state_history_stack_{fetch::chain::GENESIS_MERKLE_ROOT};
-  Hash                         current_hash_{fetch::chain::GENESIS_MERKLE_ROOT};
+  StateHashStack               state_history_stack_{fetch::chain::GetGenesisDigest()};
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/fake_storage_unit.hpp:95]: 👍

File: scripts/end_to_end_test/end_to_end_test.yaml [hunk 8]
```diff
@@ -110,6 +110,7 @@ setup_conditions:
 steps:
   - send_txs: { "name": "batch1", "amount": 10, "nodes": [0] }
   - verify_txs: { "name": "batch1", "nodes": [0] }
+  - sleep: 20
```
Comment (n-hutton) [inline scripts/end_to_end_test/end_to_end_test.yaml:113]: Is this due to inceased start up time?

**Review by n-hutton — APPROVED** [review]

File: scripts/end_to_end_test/end_to_end_test.yaml (hunk 8 above)
Comment (ejfitzgerald) [reply scripts/end_to_end_test/end_to_end_test.yaml:113]: Yes that is correct

File: libs/constellation/src/constellation.cpp (hunk 6 above)
Comment (ejfitzgerald) [reply libs/constellation/src/constellation.cpp:801]: Yes this is over kill but I want to have the shutdown exactly when when this function is called

File: libs/constellation/src/constellation.cpp (hunk 2 above)
Comment (ejfitzgerald) [reply libs/constellation/src/constellation.cpp:98]: Good spot - it was a single parameter function at one point. I think I will fix this on delta if you don't mind

**Pull request merged**